src/EditorFeatures/Core/Shared/Extensions/OperationContextExtensions.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Shared.Extensions
+{
+    internal static class OperationContextExtensions
+    {
+        /// <summary>
+        /// Wait until given document is fully loaded and then add scope
+        /// </summary>
+        public static IUIThreadOperationScope WaitForDocumentAndAddScope(
+            this IUIThreadOperationContext context, Document document, bool allowCancellation, string description)
+        {
+            return WaitAndAddScope(context, document.Project.Solution, document.Name, allowCancellation, description);
+        }
+
+        public static IUIThreadOperationScope WaitForProjectAndAddScope(
+            this IUIThreadOperationContext context, Project project, bool allowCancellation, string description)
+        {
+            return WaitAndAddScope(context, project.Solution, project.Name, allowCancellation, description);
+        }
+
+        private static IUIThreadOperationScope WaitAndAddScope(
+            this IUIThreadOperationContext context, Solution solution, string name, bool allowCancellation, string description)
+        {
+            IUIThreadOperationScope scope = null;
+
+            try
+            {
+                // partial mode is always cancellable
+                var title = string.Format(EditorFeaturesResources.Operation_is_not_ready_for_0_yet_see_task_center_for_more_detail, name);
+                scope = context.AddScope(allowCancellation: true, title);
+
+                var service = solution.Workspace.Services.GetService<ISolutionStatusService>();
+                if (service != null)
+                {
+                    // UIThreadOperation always called from UI thread.
+                    service.WaitForAsync(solution, scope.Context.UserCancellationToken).Wait(context.UserCancellationToken);
```

Comment:
referencing socpe.Context.UserCancellationToken *and* context.UserCancellationToken is extremely confusing.

Reply:
it is unclear from a design perspective what this could mean.  What does it mean to 'wait for a solution...'?  even if i expand out the terminology "wait for a solution to eb fully loaded", it still doesn't make sense.  A Roslyn Solution is immutbale.  It is already realized, and there is no concept of it being otherwise.

We could *potentially* make this a Workspace APi though.   As the Workspace is intrinsically mutable, it at least makes some amount of sense to have the concept of waitin for a point until we think no more mutations must come in for that Workspace.  

However, this is still super weird.  Even if you wait for the workspace, you still hvae your *original* solution.  So you really aren't capable of proceeding, because mutations may have happened you need to know about.

Reply:
I guess you don't know how IUIThreadOperationContext works. probably read doc comment on IUIThreadOperationContext?

or are you saying just use context.UserCancellationToken?

Reply:
I am not sure what you are asking. if you have suggestion for better wording? just propose that so that people can discuss.

Reply:
> I guess you don't know how IUIThreadOperationContext works. probably read doc comment on IUIThreadOperationContext?

I know how it works.  Which is why teh code is confusing.

Reply:
> I am not sure what you are asking. if you have suggestion for better wording? just propose that so that people can discuss.

I just did.  I'm proposing that this not be related to the Solution since it's *nonsensical* given waht a Solution represents in Roslyn.  Please see posts like https://github.com/dotnet/roslyn/pull/33779#discussion_r261787937

Reply:
I get that it is confusing for you. not sure what I can do when you are confusing. if you know 2 are same thing. what confuse you? are you saying just do context.cancellation? if that's what you want, just say, hwo about just put "context.UserCancellationToken" since scope.context and context here is same thing?

Reply:
I stil not sure what word are you proposing except you throwing it is nonsensical, strange, confusing and etc, etc.

Reply:
> I get that it is confusing for you. not sure what I can do when you are confusing. if you know 2 are same thing. what confuse you? are you saying just do context.cancellation? if that's what you want, just say, hwo about just put "context.UserCancellationToken" since scope.context and context here is same thing?

> if you know 2 are same thing. what confuse you? 

You are referencing what should be the same thing through entirely different paths of poperties/expressions.  That is fundamentally confusing.  It makes the code harder to read and understand.

The person reading the code now has to go: wait... i thought these wer ethe same... are they not?  Why is the code referencing it in different ways.

And, because there is no comment, one now has to go: "is this intentional?  would it be subtly wrong if to switch one to use the same expression as the other?"

This sort of coding is brittle and leads to difficulty with maintaining things.  it should be avoided, especially around exceptionally subtle stuff like cancellation/threading.

Reply:
> I stil not sure what word are you proposing except you throwing it is nonsensical, strange, confusing and etc, etc.

I am proposing you choose a coding pattern (your choice of which) which makes this less confusing and avoids the types of questions that the above code causes.  In general, i would always push back on *ever* seeing:

`Foo(x.y.z.CancellationToken).Wait(someOtherPath.CancellationToken)`

If you have this, it absolutely must be doucmented as to why this is appropriate.  If it's not appropriate it should be fixed up to be clearer, ideally by following the patterns used in the rest of hte IDE.

Reply:
I am not going to keep trying a different pattern to satisfy you. if you don't have any better suggestion then I dont like it. then I go with what I have.

Reply:
> The person reading the code now has to go: wait... i thought these wer ethe same... are they not? Why is the code referencing it in different ways.

so that's why I ask you whether you know about the UIThreadOperationContext stuff? since something obvious to me might not same for you?

I know SemanticModel.SyntaxTree is same tree as Document.GetSyntaxTree or node.SyntaxTree so I dont ask or say it is confusing when people use any variants of those since for me it is so obvious.

so I asked and you said you are familiar of the API so I wasn't sure what is confusing you. also if that specifically what confused you, you could just say, how about change all to context.UserCancellationToken since 2 are same without throwing all those words.

Reply:
> I am not going to keep trying a different pattern to satisfy you. if you don't have any better suggestion then "I dont like it". then I go with what I have.

I literally specified many times what would be acceptable.

Reply:
> so that's why I ask you whether you know about the UIThreadOperationContext stuff? since something obvious to me might not same for you?

I understand UIThreadOperationContext .  And it's because i understand it that *made* the code so confusing.  It woudl be akin to me seeing:

`Task.Run(() => Foo(x.y.CancellationToken), a.b.CancellationToken)`

This is *not* a coding pattern we should have in the IDE.  It's actively unnecessarily confusing.  Please fixup the code.  You could have literally fixed in in 5 seconds, rather than forcing all this back and forth.   It's completely unnecessary for you to engage in thsi behavior.

Reply:
> are same without throwing all those words.

I don't know what this means.  I'm taking the time to fully explain my concerns about this issue.  When i've been brief in the past, you've complained that i'm nto being clear.  So i'm taking a lot of time and effort to explain the problem and to ask you to fix it.  If you are confused by my request, you are welcome to come talk to me on gitter.im so we can work through it.  I've asked if you would do this in the past, and so far you have never once taken me up on just conversing in that manner.

Note: i get the sense that github communication does not work very well with you.  I'm open to trying alternative approaches (as i've said in the past).  The ball is in your court if you are open to trying that as well.

Reply:
but it is not a.b.x random thing you come up to make your opinion looks right. it is context and scope.context.. anyway, I think what you were trying to say was just use same context. right? I changed them to use same context.

Reply:
>it is context and scope.context.

That is exactly what i said.  If i saw: 

```c#
Task.Run(() => Foo(scope.Context.UserCancellationToken)).Wait(context.UserCancellationToken);
```

It would be equally confusing, and i would insist it be changed.  I'm nto sure why this is *at all* contentious Heejae.  We've seen how difficult it can be to get properly behaving Task/Async/Cancellation code over the lifetime of Roslyn.  Writing code that uses different paths like this to get at cancellation tokens and which mixes/matches that in a single-statement is just *confusing*.

At best, there are no problems with it.  At worst, tehre are problems with it, but people are very nervous to change it because it's unclear what the implications would be.

We should strive to make code as clear and sensible as possible.  it's good for PRs.  It's good for avoiding bugs.  And it's extremely good for maintaining things.